File: backend/src/main/java/br/edu/ifmg/cli/services/ScriptGenerator.java
package br.edu.ifmg.cli.services;

import java.util.regex.Pattern;
import java.util.stream.Collectors;
import br.edu.ifmg.cli.models.ast.*;

public class ScriptGenerator {

	private static final Pattern SAFE_ARGUMENT_PATTERN = Pattern.compile("^[a-zA-Z0-9._/-]+$");

	public String generate(AstNode rootNode) {
		if (rootNode == null)
			throw new IllegalArgumentException("AST não pode ser nula");
		return dispatch(rootNode);
	}

	private String dispatch(AstNode node) {
		if (node == null)
			return "";
		return switch (node.type()) {
		case AstVocabulary.Nodes.SCRIPT -> generateScript(node);
		case AstVocabulary.Nodes.COMMAND -> generateCommand(node);
		case AstVocabulary.Nodes.CONTROL -> generateControl(node);
		case AstVocabulary.Nodes.OPERATOR -> generateOperator(node);
		case AstVocabulary.Nodes.OPTION -> generateOption(node);
		case AstVocabulary.Nodes.OPERAND -> generateOperand(node);
		default -> "";
		};
	}

	private String generateControl(AstNode node) {
		if (node.controlConfig() == null)
			return "";

		var controlConfig = node.controlConfig();
		var sb = new StringBuilder(node.name());

		for (var slot : controlConfig.slots()) {
			var parameterOpt = node.getParameter(slot.key());

			if (parameterOpt.isPresent()) {
				var parameter = parameterOpt.get();
				String content = renderParameter(parameter, "\n");

				if (content.isBlank() && !slot.obligatory())
					continue;

				if (slot.breakLineBefore()) {
					if (sb.length() > 0 && sb.charAt(sb.length() - 1) != '\n') {
						sb.append("\n");
					}
				} else {
					ensureSpaceSeparator(sb);
				}

				if (slot.syntaxPrefix() != null) {
					sb.append(slot.syntaxPrefix());
				}

				if (parameter.isContainer()) {
					sb.append("\n").append(indent(content));
				} else {
					ensureSpaceSeparator(sb);
					sb.append(content);
				}
			}
		}

		if (controlConfig.syntaxEnd() != null) {
			if (sb.length() > 0 && sb.charAt(sb.length() - 1) != '\n') {
				sb.append("\n");
			}
			sb.append(controlConfig.syntaxEnd());
		}

		return sb.toString();
	}

	private void ensureSpaceSeparator(StringBuilder sb) {
		if (sb.length() > 0) {
			char lastChar = sb.charAt(sb.length() - 1);
			if (lastChar != '\n' && lastChar != ' ') {
				sb.append(" ");
			}
		}
	}

	private String generateScript(AstNode node) {
		return node.parameters().stream().map(p -> renderParameter(p, "\n")).filter(s -> !s.isBlank())
				.collect(Collectors.joining("\n"));
	}

	private String generateCommand(AstNode node) {
		var sb = new StringBuilder(node.name());

		node.getParameter(AstVocabulary.Keys.OPTIONS).ifPresent(p -> {
			String val = renderParameter(p, " ");
			if (!val.isBlank())
				sb.append(" ").append(val);
		});

		node.getParameter(AstVocabulary.Keys.OPERANDS).ifPresent(p -> {
			String val = renderParameter(p, " ");
			if (!val.isBlank())
				sb.append(" ").append(val);
		});

		return sb.toString();
	}

    private String generateOperator(AstNode node) {
        if (node.operatorConfig() == null)
            return "";

        var sb = new StringBuilder();
        var config = node.operatorConfig();

        for (var slot : config.slots()) {
            node.getParameter(slot.key()).ifPresent(parameter -> {
                String content;
                if (parameter.isContainer()) {
                    content = renderChildren(parameter, "\n");
                } else {
                    String rawValue = parameter.value();
                    if (rawValue == null || rawValue.isBlank()) return;
                    content = quoteArgumentIfUnsafe(rawValue);
                }

                if (content.isBlank())
                    return;

                if (sb.length() > 0)
                    sb.append(" ");

                if (slot.symbol() != null) {
                    if (AstVocabulary.Values.PLACEMENT_BEFORE.equals(slot.symbolPlacement())) {
                        sb.append(slot.symbol()).append(" ").append(content);
                    } else {
                        sb.append(content).append(" ").append(slot.symbol());
                    }
                } else {
                    sb.append(content);
                }
            });
        }
        return sb.toString();
    }

	private String generateOption(AstNode node) {
		String flag = node.getParameter(AstVocabulary.Keys.FLAG).map(AstParameter::value).orElse("");
		String value = node.getParameter(AstVocabulary.Keys.VALUE).map(AstParameter::value).orElse("");
		var sb = new StringBuilder();
		if (!flag.isBlank()) {
			sb.append(flag);
			if (!value.isBlank())
				sb.append(" ").append(quoteArgumentIfUnsafe(value));
		}
		return sb.toString();
	}

	private String generateOperand(AstNode node) {
		return node.getParameter(AstVocabulary.Keys.VALUE).map(AstParameter::value).map(this::quoteArgumentIfUnsafe)
				.orElse("");
	}

	private String renderParameter(AstParameter parameter, String separator) {
		if (parameter.isContainer()) {
			return renderChildren(parameter, separator);
		}
		return parameter.value();
	}

	private String renderChildren(AstParameter parameter, String separator) {
		return parameter.children().stream().map(this::dispatch).filter(rendered -> !rendered.isBlank())
				.collect(Collectors.joining(separator));
	}

	private String quoteArgumentIfUnsafe(String rawArgument) {
		if (rawArgument == null || rawArgument.isEmpty())
			return "''";
		if (SAFE_ARGUMENT_PATTERN.matcher(rawArgument).matches())
			return rawArgument;
		return "'" + rawArgument.replace("'", "'\\''") + "'";
	}

	private String indent(String code) {
		return code.lines().map(line -> "  " + line).collect(Collectors.joining("\n"));
	}
}

File: frontend/src/core/shellblocks/generation/scriptGenerator.ts
import { ASTNode, ASTParameter } from "../types/ast"; // Ajuste o caminho conforme sua árvore

const SAFE_ARGUMENT_PATTERN = /^[a-zA-Z0-9._/-]+$/;

export function generateShellScript(rootNode: ASTNode | undefined | null): string {
    if (!rootNode) throw new Error("AST não pode ser nula");
    return dispatch(rootNode);
}

function dispatch(node: ASTNode | undefined): string {
    if (!node) return "";

    switch (node.type) {
        case "script": return generateScript(node);
        case "command": return generateCommand(node);
        case "control": return generateControl(node);
        case "operator": return generateOperator(node);
        case "option": return generateOption(node);
        case "operand": return generateOperand(node);
        default: return "";
    }
}

function generateScript(node: ASTNode): string {
    if (!node.parameters) return "";
    return node.parameters
        .map((p) => renderParameter(p, "\n"))
        .filter((s) => s.trim() !== "")
        .join("\n");
}

function generateCommand(node: ASTNode): string {
    let sb = node.name || "";

    const optionsParam = getParameter(node, "options");
    if (optionsParam) {
        const val = renderParameter(optionsParam, " ");
        if (val.trim() !== "") sb += ` ${val}`;
    }

    const operandsParam = getParameter(node, "operands");
    if (operandsParam) {
        const val = renderParameter(operandsParam, " ");
        if (val.trim() !== "") sb += ` ${val}`;
    }

    return sb;
}

function generateControl(node: ASTNode): string {
    if (!node.controlConfig) return "";

    let sb = node.name || "";

    for (const slot of node.controlConfig.slots) {
        const parameter = getParameter(node, slot.key);

        if (parameter) {
            const content = renderParameter(parameter, "\n");

            if (content.trim() === "" && !slot.obligatory) continue;

            sb = ensureSpaceSeparator(sb);

            if (slot.syntaxPrefix) {
                sb += slot.syntaxPrefix;
            }

            if (isContainer(parameter)) {
                sb += `\n${indent(content)}`;
            } else {
                sb = ensureSpaceSeparator(sb);
                sb += content;
            }
        }
    }

    if (node.controlConfig.syntaxEnd) {
        if (sb.length > 0 && !sb.endsWith("\n")) sb += "\n";
        sb += node.controlConfig.syntaxEnd;
    }

    return sb;
}

function generateOperator(node: ASTNode): string {
    if (!node.operatorConfig) return "";

    let sb = "";

    for (const slot of node.operatorConfig.slots) {
        const parameter = getParameter(node, slot.key);

        if (parameter) {
            let content = "";

            if (isContainer(parameter)) {
                content = renderChildren(parameter, "\n");
            } else {
                const rawValue = parameter.value;
                if (!rawValue || rawValue.trim() === "") continue;
                content = quoteArgumentIfUnsafe(rawValue);
            }

            if (content.trim() === "") continue;

            if (slot.breakLineBefore) {
                if (sb.length > 0 && !sb.endsWith("\n")) sb += "\n";
            } else if (sb.length > 0) {
                sb += " ";
            }

            if (slot.symbol) {
                if (slot.symbolPlacement === "before") {
                    sb += `${slot.symbol} ${content}`;
                } else {
                    sb += `${content} ${slot.symbol}`;
                }
            } else {
                sb += content;
            }
        }
    }
    return sb;
}

function generateOption(node: ASTNode): string {
    const flag = getParameterValue(node, "flag");
    const value = getParameterValue(node, "value");

    let sb = "";
    if (flag.trim() !== "") {
        sb += flag;
        if (value.trim() !== "") {
            sb += ` ${quoteArgumentIfUnsafe(value)}`;
        }
    }
    return sb;
}

function generateOperand(node: ASTNode): string {
    const value = getParameterValue(node, "value");
    return value ? quoteArgumentIfUnsafe(value) : "";
}

// --- Funções Auxiliares ---

function getParameter(node: ASTNode, key: string): ASTParameter | undefined {
    return node.parameters?.find((p) => p.key === key);
}

function getParameterValue(node: ASTNode, key: string): string {
    return getParameter(node, key)?.value || "";
}

function isContainer(parameter: ASTParameter): boolean {
    return !!parameter.children && parameter.children.length > 0;
}

function renderParameter(parameter: ASTParameter, separator: string): string {
    if (isContainer(parameter)) {
        return renderChildren(parameter, separator);
    }
    return parameter.value || "";
}

function renderChildren(parameter: ASTParameter, separator: string): string {
    if (!parameter.children) return "";
    return parameter.children
        .map((child) => dispatch(child))
        .filter((rendered) => rendered.trim() !== "")
        .join(separator);
}

function ensureSpaceSeparator(str: string): string {
    if (str.length > 0) {
        const lastChar = str.slice(-1);
        if (lastChar !== "\n" && lastChar !== " ") {
            return str + " ";
        }
    }
    return str;
}

function quoteArgumentIfUnsafe(rawArgument: string): string {
    if (!rawArgument || rawArgument.length === 0) return "''";
    if (SAFE_ARGUMENT_PATTERN.test(rawArgument)) return rawArgument;
    return `'${rawArgument.replace(/'/g, "'\\''")}'`;
}

function indent(code: string): string {
    return code
        .split("\n")
        .map((line) => `  ${line}`)
        .join("\n");
}
